File: practica parte 1/Ejercicio 5: Clase genérica Pila/pila.py
class Pila:
    def __init__(self):
        self.elementos = []
    
    def apilar(self, item):
        self.elementos.append(item)
    
    def desapilar(self):
        if not self.esta_vacia():
            return self.elementos.pop()
        return None
    
    def mostrar(self):
        print("Contenido de la pila (LIFO):")
        for item in reversed(self.elementos):
            print(item)
    
    def esta_vacia(self):
        return len(self.elementos) == 0

# Prueba con diferentes tipos
pila_enteros = Pila()
pila_enteros.apilar(10)
pila_enteros.apilar(20)
pila_enteros.apilar(30)
pila_enteros.mostrar()
print("Desapilado:", pila_enteros.desapilar())

pila_strings = Pila()
pila_strings.apilar("Uno")
pila_strings.apilar("Dos")
pila_strings.apilar("Tres")
pila_strings.mostrar()

File: practica parte 1/Ejercicio 5: Clase genérica Pila/src/pila.java
import java.util.ArrayList;
import java.util.List;

class Pila<T> {
    private List<T> elementos;
    
    public Pila() {
        elementos = new ArrayList<>();
    }
    
    public void apilar(T item) {
        elementos.add(item);
    }
    
    public T desapilar() {
        if (!estaVacia()) {
            return elementos.remove(elementos.size() - 1);
        }
        return null;
    }
    
    public void mostrar() {
        System.out.println("Contenido de la pila (LIFO):");
        for (int i = elementos.size() - 1; i >= 0; i--) {
            System.out.println(elementos.get(i));
        }
    }
    
    public boolean estaVacia() {
        return elementos.isEmpty();
    }
}

public class MainPila {
    public static void main(String[] args) {
        Pila<Integer> pilaEnteros = new Pila<>();
        pilaEnteros.apilar(10);
        pilaEnteros.apilar(20);
        pilaEnteros.apilar(30);
        pilaEnteros.mostrar();
        System.out.println("Desapilado: " + pilaEnteros.desapilar());
        
        Pila<String> pilaStrings = new Pila<>();
        pilaStrings.apilar("Uno");
        pilaStrings.apilar("Dos");
        pilaStrings.apilar("Tres");
        pilaStrings.mostrar();
    }
}